File: src/main/java/org/coinen/reactive/pacman/repository/support/InMemoryExtrasRepository.java
package org.coinen.reactive.pacman.repository.support;

import org.coinen.reactive.pacman.repository.ExtrasRepository;
import org.roaringbitmap.PeekableIntIterator;
import org.roaringbitmap.RoaringBitmap;

public class InMemoryExtrasRepository implements ExtrasRepository {

    final int tileSize = 100;
    final int mapWidth = 60;
    final int mapHeight = 60;
    final int offset = 11;

    final RoaringBitmap bitmap = new RoaringBitmap();

    @Override
    public int collideExtra(float x, float y) {
        var i = Math.round(x / tileSize);
        var j = Math.round(y / tileSize);

        var flattenPosition = i + j * mapWidth;

        if (bitmap.checkedRemove(flattenPosition)) {
            return flattenPosition;
        } else if(bitmap.checkedRemove(-flattenPosition)) {
            return -flattenPosition;
        }

        return 0;
    }

    @Override
    public int createExtra(int size) {
        var nextPosition = 0;
        do {
            nextPosition = randomPosition(mapWidth, mapHeight, offset);
        } while(bitmap.contains(nextPosition) || bitmap.contains(-nextPosition));

        var powerupCount = 0;

        PeekableIntIterator iterator = bitmap.getIntIterator();

        while (iterator.hasNext()) {
            if (iterator.next() < 0) {
                powerupCount++;
            }
        }

        var powerupAllowed = false;
        var totalSpace =
            (mapWidth - 2 * offset + 1) * (mapHeight - 2 * offset + 1);
        if (size <= 2) {
            powerupAllowed = powerupCount < Math.ceil(totalSpace / 360);
        }
        else if (size <= 4) {
            powerupAllowed = powerupCount < Math.ceil(totalSpace / 900);
        }
        else {
            powerupAllowed = powerupCount == 0 && Math.random() < 0.02;
        }

        if (powerupAllowed) {
            bitmap.add(-nextPosition);
            return -nextPosition;
        }
        else {
            bitmap.add(nextPosition);
            return nextPosition;
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public Iterable<Integer> finaAll() {
        return bitmap;
    }

    @Override
    public void saveAll(int[] extras) {
        bitmap.add(extras);
    }


    static int randomPosition(int width, int height, int offset) {
        return (int) (Math.floor(Math.random() * (width - 2 * offset + 1) + offset) +  Math.floor(Math.random() * (height - 2 * offset + 1) + offset) * height);
    }
}

File: game-server/src/main/typescript/lib/repository/support/InMemoryExtrasRepository.ts
import ExtrasRepository from '../ExtrasRepository';

export default class InMemoryExtrasRepository implements ExtrasRepository {

    tileSize = 100;
    mapHeight = 60;
    mapWidth = 60;
    offset = 11;

    bitmap: Set<number> = new Set();

    collideExtra(x: number, y: number): number {
        let i = Math.round(x / this.tileSize);
        let j = Math.round(y / this.tileSize);

        const flattenPosition = i + j * this.mapWidth;
        if (this.bitmap.has(flattenPosition)) {
            this.bitmap.delete(flattenPosition);
            return flattenPosition;
        } else if (this.bitmap.has(-flattenPosition)) {
            this.bitmap.delete(flattenPosition);
            return -flattenPosition;
        }

        return 0;
    }


    
    createExtra(size: number): number {
        let nextPosition = 0;
        do {
            nextPosition = InMemoryExtrasRepository.randomPosition(this.mapWidth, this.mapHeight, this.offset);
        } while (this.bitmap.has(nextPosition) || this.bitmap.has(-nextPosition));

        let powerupCount = 0;
        this.bitmap.forEach(el => Math.sign(el) < 0 ? powerupCount++ : {});

        var powerupAllowed = false;
        var totalSpace =
            (this.mapWidth - 2 * this.offset + 1) * (this.mapHeight - 2 * this.offset + 1);
        if (size <= 2) {
            powerupAllowed = powerupCount < Math.ceil(totalSpace / 360);
        }
        else if (size <= 4) {
            powerupAllowed = powerupCount < Math.ceil(totalSpace / 900);
        }
        else {
            powerupAllowed = powerupCount == 0 && Math.random() < 0.02;
        }

        if (powerupAllowed) {
            this.bitmap.add(-nextPosition);
            return -nextPosition;
        }
        else {
            this.bitmap.add(nextPosition);
            return nextPosition;
        }
    }

    findAll(): Set<number> {
        return this.bitmap;
    }

    saveAll(extras: number[]): void {
        extras.forEach(extra => {
            this.bitmap.add(extra);
        });
        console.log('extras', this.bitmap);
    }


    static randomPosition(width: number, height: number, offset: number): number {
        return (Math.floor(Math.random() * (width - 2 * offset + 1) + offset) +  Math.floor(Math.random() * (height - 2 * offset + 1) + offset) * height);
    }
}
